src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java

```diff
@@ -278,93 +278,100 @@ private InetSocketAddress createInetSocketAddress(String host, int port) {
                     socket.connect(endpoint, connectTimeout);
-            }
+                }
 
-            // continue (but ignore connectTimeout)
+                // continue (but ignore connectTimeout)
 
-            if (socket == null) {
-                if (debug) {
-                    System.err.println("Connection: creating socket");
+                if (socket == null) {
+                    if (debug) {
+                        System.err.println("Connection: creating socket");
+                    }
+                    // connected socket
+                    socket = new Socket(host, port);
                 }
-                // connected socket
-                socket = new Socket(host, port);
             }
-        }
 
-        // For LDAP connect timeouts on LDAP over SSL connections must treat
-        // the SSL handshake following socket connection as part of the timeout.
-        // So explicitly set a socket read timeout, trigger the SSL handshake,
-        // then reset the timeout.
-        if (socket instanceof SSLSocket) {
-            SSLSocket sslSocket = (SSLSocket) socket;
-            if (!IS_HOSTNAME_VERIFICATION_DISABLED) {
-                SSLParameters param = sslSocket.getSSLParameters();
-                param.setEndpointIdentificationAlgorithm("LDAPS");
-                sslSocket.setSSLParameters(param);
+            // For LDAP connect timeouts on LDAP over SSL connections must treat
+            // the SSL handshake following socket connection as part of the timeout.
+            // So explicitly set a socket read timeout, trigger the SSL handshake,
+            // then reset the timeout.
+            if (socket instanceof SSLSocket) {
+                SSLSocket sslSocket = (SSLSocket) socket;
+                if (!IS_HOSTNAME_VERIFICATION_DISABLED) {
+                    SSLParameters param = sslSocket.getSSLParameters();
+                    param.setEndpointIdentificationAlgorithm("LDAPS");
+                    sslSocket.setSSLParameters(param);
+                }
+                setHandshakeCompletedListener(sslSocket);
+                if (connectTimeout > 0) {
+                    int socketTimeout = sslSocket.getSoTimeout();
+                    sslSocket.setSoTimeout(connectTimeout); // reuse full timeout value
+                    sslSocket.startHandshake();
+                    sslSocket.setSoTimeout(socketTimeout);
+                }
             }
-            setHandshakeCompletedListener(sslSocket);
-            if (connectTimeout > 0) {
-                int socketTimeout = sslSocket.getSoTimeout();
-                sslSocket.setSoTimeout(connectTimeout); // reuse full timeout value
-                sslSocket.startHandshake();
-                sslSocket.setSoTimeout(socketTimeout);
+        } catch (Exception e) {
+            // 8314063 the socket is not closed after the failure of handshake
+            if (socket != null && !socket.isClosed()) {
+                socket.close();
```

Comment:
a closeConnectionSocket (IIRC) was introduced in previous change which handled any IOExceptions on the close. This could be used here.
But some how that seems to have been renamed !!

709     private void closeOpenedSocket() {
 710         try {
 711             sock.close();
 712         } catch (IOException ioEx) {
 713             if (debug) {
 714                 System.err.println("Connection.closeConnectionSocket: Socket close problem: " + ioEx);
 715                 System.err.println("Socket isClosed: " + sock.isClosed());
 716             }
 717         }
 718     }